Incompatible with Pyro

Deciding to document this as a few people are reporting this now. I do not have a Pyro license so this makes testing difficult, I'll probably bother @HoratioGamer about testing possible solutions, though I believe this has to be fixed on @GiantNuker's end. This same issue also affects Optifine standalone.

Note that in the first log `ref4063632994977090211.tmp` is Pyro's mixin refmap. For some reason this is included in the mods folder as well, so you can see a fairly harmless stacktrace failing to open it as a zip file. Annoyingly enough, Pyro decides to use temporary files for it's mixin refmap, which are created and deleted on game open / close / crash.

Right before the crash stacktrace is a stacktrace with this
```
[10:01:18] [Client thread/WARN] [mixin]: @Redirect conflict. Skipping mixins.pyroclient.json:RenderGlobalMixin->@Redirect::redirect$checkFreecamChunkUpdates$zga000(Lnet/minecraft/client/renderer/ViewFrustum;DD)V with priority 1000, already redirected by mixins.kami.json:render.MixinRenderGlobal->@Redirect::redirect$updateSetupTerrain$zce000(Lnet/minecraft/client/renderer/ViewFrustum;DD)V with priority 1000
```
The most recent change made on our end was 20 days ago, at the time of writing, [here](https://github.com/kami-blue/client/commit/b23e73968dd9637054b8cd0e0524e763f22e431a#diff-ac227dbf883c17816987174693c527f622f2dd70774e46a93440c961459a05af).

History of said file [here](https://github.com/kami-blue/client/commits/master/src/main/java/me/zeroeightsix/kami/mixin/client/render/MixinRenderGlobal.java).

Funnily enough Mixin is saying it's caused by redirecting `setupTerrain`, which has been in place for a bit longer than the current changes. I will need an exact version hash of when this breaks from @HoratioGamer otherwise this makes troubleshooting harder.

The comment above said redirect already explains why it is necessary.
https://github.com/kami-blue/client/blob/b23e73968dd9637054b8cd0e0524e763f22e431a/src/main/java/me/zeroeightsix/kami/mixin/client/render/MixinRenderGlobal.java#L64-L67

This is most likely not the cause of the issue, so, continuing forward we have:
```
[10:01:18] [Client thread/WARN] [mixin]: Method overwrite conflict for getEntityOutlineShader in mixins.pyroclient.json:RenderGlobalAccessor, previously written by me.zeroeightsix.kami.mixin.client.accessor.render.AccessorRenderGlobal. Skipping method.
```

We certainly don't do any overwriting as far as I'm aware. This is just a warning, it just means our accessor will work while Pyro's presumably won't.

Now we have an actual stacktrace, 
```
[10:01:18] [Client thread/INFO] [STDERR]: [org.spongepowered.asm.mixin.transformer.MixinTransformer:transformClassBytes:519]: org.spongepowered.asm.mixin.injection.throwables.InjectionError: Critical injection failure: Redirector redirect$checkFreecamChunkUpdates$zga000(Lnet/minecraft/client/renderer/ViewFrustum;DD)V in mixins.pyroclient.json:RenderGlobalMixin failed injection check, (0/1) succeeded. Using refmap file:/tmp/ref4063632994977090211.tmp
```

Which doesn't look promising. It looks like Pyro's redirect is failing due to ours, which is necessary as explained above. They probably have a redirect for the same reason? Not sure what could be done about that.

This stacktrace is also referenced at the end of the *actual* stacktrace here
```
java.lang.NoClassDefFoundError: net/minecraft/client/renderer/RenderGlobal
	at net.minecraftforge.client.ForgeHooksClient.shouldUseVanillaReloadableListener(ForgeHooksClient.java:951)
	at net.minecraft.client.resources.SimpleReloadableResourceManager.func_110544_b(SimpleReloadableResourceManager.java:131)
	at net.minecraft.client.resources.SimpleReloadableResourceManager.func_110541_a(SimpleReloadableResourceManager.java:112)
	at net.minecraft.client.Minecraft.func_110436_a(Minecraft.java:808)
	at net.minecraftforge.fml.client.FMLClientHandler.beginMinecraftLoading(FMLClientHandler.java:247)
	at net.minecraft.client.Minecraft.func_71384_a(Minecraft.java:467)
	at net.minecraft.client.Minecraft.func_99999_d(Minecraft.java:378)
	at net.minecraft.client.main.Main.main(SourceFile:123)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at net.minecraft.launchwrapper.Launch.launch(Launch.java:135)
	at net.minecraft.launchwrapper.Launch.main(Launch.java:28)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.multimc.onesix.OneSixLauncher.launchWithMainClass(OneSixLauncher.java:196)
	at org.multimc.onesix.OneSixLauncher.launch(OneSixLauncher.java:231)
	at org.multimc.EntryPoint.listen(EntryPoint.java:143)
	at org.multimc.EntryPoint.main(EntryPoint.java:34)
Caused by: java.lang.ClassNotFoundException: net.minecraft.client.renderer.RenderGlobal
	at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:191)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	... 22 more
Caused by: org.spongepowered.asm.mixin.transformer.throwables.MixinTransformerError: An unexpected critical error was encountered
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClassBytes(MixinTransformer.java:521)
	at org.spongepowered.asm.mixin.transformer.Proxy.transform(Proxy.java:72)
	at net.minecraft.launchwrapper.LaunchClassLoader.runTransformers(LaunchClassLoader.java:279)
	at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:176)
	... 24 more
Caused by: org.spongepowered.asm.mixin.injection.throwables.InjectionError: Critical injection failure: Redirector redirect$checkFreecamChunkUpdates$zga000(Lnet/minecraft/client/renderer/ViewFrustum;DD)V in mixins.pyroclient.json:RenderGlobalMixin failed injection check, (0/1) succeeded. Using refmap file:/tmp/ref4063632994977090211.tmp
	at org.spongepowered.asm.mixin.injection.struct.InjectionInfo.postInject(InjectionInfo.java:290)
	at org.spongepowered.asm.mixin.transformer.MixinTargetContext.applyInjections(MixinTargetContext.java:1203)
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.applyInjections(MixinApplicatorStandard.java:938)
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.applyMixin(MixinApplicatorStandard.java:322)
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.apply(MixinApplicatorStandard.java:280)
	at org.spongepowered.asm.mixin.transformer.TargetClassContext.applyMixins(TargetClassContext.java:353)
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.apply(MixinTransformer.java:724)
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.applyMixins(MixinTransformer.java:703)
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClassBytes(MixinTransformer.java:509)
	... 27 more
```

Which is fairly interesting. I'm not quite sure why failing to redirect a method will cause `java.lang.NoClassDefFoundError: net/minecraft/client/renderer/RenderGlobal`. Tried troubleshooting this before, in tickets 109 and 201 (201 has been left open until this is resolved).

I'm not quite sure what to do from here, the log can be found [here](https://paste.ee/p/W9gnG) for more investigation. 

Log running on the first version that breaks (2.01.01-baf7588e) can be found [here](https://cdn.discordapp.com/attachments/804377170839207986/804392283142619137/debug-2.01.01-baf7588e.log). Knowing the version isn't very helpful, as that version was a 16k line diff, since it was the GUI update

--- a/src/main/java/org/kamiblue/client/mixin/client/render/MixinRenderGlobal.java
+++ b/src/main/java/org/kamiblue/client/mixin/client/render/MixinRenderGlobal.java
@@ -1,36 +1,21 @@
 package org.kamiblue.client.mixin.client.render;
 
-import net.minecraft.client.Minecraft;
-import net.minecraft.client.entity.EntityPlayerSP;
 import net.minecraft.client.renderer.RenderGlobal;
-import net.minecraft.client.renderer.ViewFrustum;
-import net.minecraft.client.renderer.chunk.RenderChunk;
 import net.minecraft.entity.player.EntityPlayer;
-import net.minecraft.util.EnumFacing;
 import net.minecraft.util.math.BlockPos;
-import net.minecraft.util.math.MathHelper;
 import net.minecraft.util.math.RayTraceResult;
 import org.kamiblue.client.event.KamiEventBus;
 import org.kamiblue.client.event.events.BlockBreakEvent;
-import org.kamiblue.client.mixin.client.accessor.render.AccessorViewFrustum;
 import org.kamiblue.client.module.modules.player.Freecam;
 import org.kamiblue.client.module.modules.render.SelectionHighlight;
-import org.kamiblue.client.util.Wrapper;
 import org.spongepowered.asm.mixin.Mixin;
-import org.spongepowered.asm.mixin.Shadow;
 import org.spongepowered.asm.mixin.injection.At;
 import org.spongepowered.asm.mixin.injection.Inject;
-import org.spongepowered.asm.mixin.injection.Redirect;
+import org.spongepowered.asm.mixin.injection.ModifyVariable;
 import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
 
 @Mixin(RenderGlobal.class)
 public abstract class MixinRenderGlobal {
-
-    @Shadow private int renderDistanceChunks;
-    @Shadow private ViewFrustum viewFrustum;
-
-    private final Minecraft mc = Minecraft.getMinecraft();
-
     @Inject(method = "drawSelectionBox", at = @At("HEAD"), cancellable = true)
     public void drawSelectionBox(EntityPlayer player, RayTraceResult movingObjectPositionIn, int execute, float partialTicks, CallbackInfo ci) {
         if (SelectionHighlight.INSTANCE.isEnabled() && SelectionHighlight.INSTANCE.getBlock().getValue()) {
@@ -44,38 +29,12 @@ public abstract class MixinRenderGlobal {
         KamiEventBus.INSTANCE.post(event);
     }
 
-    // Can't use @ModifyVariable here because it crashes outside of a dev env with Optifine
-    @Redirect(method = "setupTerrain", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/renderer/RenderGlobal;getRenderChunkOffset(Lnet/minecraft/util/math/BlockPos;Lnet/minecraft/client/renderer/chunk/RenderChunk;Lnet/minecraft/util/EnumFacing;)Lnet/minecraft/client/renderer/chunk/RenderChunk;"))
-    public RenderChunk renderChunkOffset(RenderGlobal renderGlobal, BlockPos playerPos, RenderChunk renderChunkBase, EnumFacing facing) {
+    @ModifyVariable(method = "setupTerrain", at = @At(value = "STORE", ordinal = 0), ordinal = 1)
+    public BlockPos setupTerrainStoreFlooredChunkPosition(BlockPos playerPos) {
         if (Freecam.INSTANCE.isEnabled()) {
             playerPos = Freecam.getRenderChunkOffset(playerPos);
         }
 
-        // Can't use a @Shadow of getRenderChunkOffset because it crashes outside of a dev env with Optifine
-        BlockPos blockpos = renderChunkBase.getBlockPosOffset16(facing);
-
-        if (MathHelper.abs(playerPos.getX() - blockpos.getX()) > this.renderDistanceChunks * 16) {
-            return null;
-        } else if (blockpos.getY() >= 0 && blockpos.getY() < 256) {
-            return MathHelper.abs(playerPos.getZ() - blockpos.getZ()) > this.renderDistanceChunks * 16 ? null : ((AccessorViewFrustum) this.viewFrustum).invokeGetRenderChunk(blockpos);
-        } else {
-            return null;
-        }
-    }
-
-    /*
-     * updateChunkPositions loadRenderers as well, but as long as you don't change your renderDistance in Freecam loadRenderers won't be called
-     * One could add the same redirect for loadRenderers if needed
-     */
-    @Redirect(method = "setupTerrain", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/renderer/ViewFrustum;updateChunkPositions(DD)V"))
-    public void updateSetupTerrain(ViewFrustum viewFrustum, double viewEntityX, double viewEntityZ) {
-        EntityPlayerSP player = Wrapper.getPlayer();
-
-        if (Freecam.INSTANCE.isEnabled() && player != null) {
-            viewEntityX = player.posX;
-            viewEntityZ = player.posZ;
-        }
-
-        viewFrustum.updateChunkPositions(viewEntityX, viewEntityZ);
+        return playerPos;
     }
 }

--- a/src/main/java/org/kamiblue/client/mixin/client/render/MixinRenderPlayer.java
+++ b/src/main/java/org/kamiblue/client/mixin/client/render/MixinRenderPlayer.java
@@ -1,25 +1,29 @@
 package org.kamiblue.client.mixin.client.render;
 
 import net.minecraft.client.entity.AbstractClientPlayer;
+import net.minecraft.client.model.ModelBase;
 import net.minecraft.client.renderer.GlStateManager;
+import net.minecraft.client.renderer.entity.RenderLivingBase;
 import net.minecraft.client.renderer.entity.RenderManager;
 import net.minecraft.client.renderer.entity.RenderPlayer;
-import net.minecraft.entity.Entity;
 import net.minecraft.util.math.Vec3d;
 import org.kamiblue.client.module.modules.movement.ElytraFlight;
+import org.kamiblue.client.module.modules.player.Freecam;
 import org.kamiblue.client.util.Wrapper;
-import org.objectweb.asm.Opcodes;
 import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
 import org.spongepowered.asm.mixin.injection.At;
 import org.spongepowered.asm.mixin.injection.Inject;
-import org.spongepowered.asm.mixin.injection.Redirect;
 import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
 
-/**
- * Created by 086 on 19/12/2017.
- */
 @Mixin(RenderPlayer.class)
-public class MixinRenderPlayer {
+public abstract class MixinRenderPlayer extends RenderLivingBase<AbstractClientPlayer> {
+
+    public MixinRenderPlayer(RenderManager renderManagerIn, ModelBase modelBaseIn, float shadowSizeIn) {
+        super(renderManagerIn, modelBaseIn, shadowSizeIn);
+    }
+
+    @Shadow protected abstract void setModelVisibilities(AbstractClientPlayer clientPlayer);
 
     @Inject(method = "applyRotations", at = @At("RETURN"))
     protected void applyRotations(AbstractClientPlayer entityLiving, float ageInTicks, float rotationYaw, float partialTicks, CallbackInfo ci) {
@@ -36,9 +40,20 @@ public class MixinRenderPlayer {
         }
     }
 
-    // Redirect it to the original player so the original player can be renderer correctly
-    @Redirect(method = "doRender", at = @At(value = "FIELD", target = "Lnet/minecraft/client/renderer/entity/RenderManager;renderViewEntity:Lnet/minecraft/entity/Entity;", opcode = Opcodes.GETFIELD))
-    public Entity getRenderViewEntity(RenderManager renderManager) {
-        return Wrapper.getPlayer();
+    // Force it to render the original player in Freecam
+    @Inject(method = "doRender", at = @At(value = "FIELD", target = "Lnet/minecraft/client/renderer/entity/RenderManager;renderViewEntity:Lnet/minecraft/entity/Entity;"))
+    public void doRenderGetRenderViewEntity(AbstractClientPlayer entity, double x, double y, double z, float entityYaw, float partialTicks, CallbackInfo ci) {
+        if (Freecam.INSTANCE.isEnabled() && Wrapper.getMinecraft().getRenderViewEntity() != entity) {
+            double renderY = y;
+
+            if (entity.isSneaking()) {
+                renderY = y - 0.125D;
+            }
+
+            this.setModelVisibilities(entity);
+            GlStateManager.enableBlendProfile(GlStateManager.Profile.PLAYER_SKIN);
+            super.doRender(entity, x, renderY, z, entityYaw, partialTicks);
+            GlStateManager.disableBlendProfile(GlStateManager.Profile.PLAYER_SKIN);
+        }
     }
 }

--- a/src/main/java/org/kamiblue/client/mixin/client/render/MixinViewFrustum.java
+++ b/src/main/java/org/kamiblue/client/mixin/client/render/MixinViewFrustum.java
@@ -1,0 +1,53 @@
+package org.kamiblue.client.mixin.client.render;
+
+import net.minecraft.client.entity.EntityPlayerSP;
+import net.minecraft.client.renderer.ViewFrustum;
+import net.minecraft.client.renderer.chunk.RenderChunk;
+import net.minecraft.util.math.MathHelper;
+import org.kamiblue.client.module.modules.player.Freecam;
+import org.kamiblue.client.util.Wrapper;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+// TODO: Find a working fix of this for optifine
+@Mixin(ViewFrustum.class)
+public abstract class MixinViewFrustum {
+    @Shadow public RenderChunk[] renderChunks;
+    @Shadow protected int countChunksX;
+    @Shadow protected int countChunksY;
+    @Shadow protected int countChunksZ;
+
+    @Shadow protected abstract int getBaseCoordinate(int p_178157_1_, int p_178157_2_, int p_178157_3_);
+
+    @Inject(method = "updateChunkPositions", at = @At("HEAD"), cancellable = true)
+    public void updateChunkPositionsHead(double viewEntityX, double viewEntityZ, CallbackInfo ci) {
+        if (Freecam.INSTANCE.isDisabled()) return;
+
+        EntityPlayerSP player = Wrapper.getPlayer();
+        if (player == null) return;
+
+        int centerX = MathHelper.floor(player.posX) - 8;
+        int centerZ = MathHelper.floor(player.posZ) - 8;
+
+        int multipliedCountX = this.countChunksX * 16;
+
+        for (int x = 0; x < this.countChunksX; ++x) {
+            int posX = this.getBaseCoordinate(centerX, multipliedCountX, x);
+
+            for (int z = 0; z < this.countChunksZ; ++z) {
+                int poxZ = this.getBaseCoordinate(centerZ, multipliedCountX, z);
+
+                for (int y = 0; y < this.countChunksY; ++y) {
+                    int poxY = y * 16;
+                    RenderChunk renderchunk = this.renderChunks[(z * this.countChunksY + y) * this.countChunksX + x];
+                    renderchunk.setPosition(posX, poxY, poxZ);
+                }
+            }
+        }
+
+        ci.cancel();
+    }
+}

--- a/src/main/resources/mixins.kami.json
+++ b/src/main/resources/mixins.kami.json
@@ -72,6 +72,7 @@
     "world.MixinBlockLiquid",
     "world.MixinBlockSoulSand",
     "world.MixinBlockWeb",
+    "render.MixinViewFrustum",
     "world.MixinWorld"
   ]
 }

--- a/src/main/resources/mixins.kami.json
+++ b/src/main/resources/mixins.kami.json
@@ -68,11 +68,11 @@
     "render.MixinRenderPlayer",
     "render.MixinTileEntityRendererDispatcher",
     "render.MixinTileEntitySignRenderer",
+    "render.MixinViewFrustum",
     "render.MixinVisGraph",
     "world.MixinBlockLiquid",
     "world.MixinBlockSoulSand",
     "world.MixinBlockWeb",
-    "render.MixinViewFrustum",
     "world.MixinWorld"
   ]
 }

